contracts/libraries/SwapToRatio.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.7.0;
+
+import '@uniswap/v3-core/contracts/libraries/TickMath.sol';
+import '@uniswap/v3-core/contracts/libraries/SqrtPriceMath.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import './PoolTicksLibrary.sol';
+import 'hardhat/console.sol';
+
+library SwapToRatio {
+    using PoolTicksLibrary for IUniswapV3Pool;
+
+    struct PoolParams {
+        uint160 sqrtRatioX96;
+        uint128 liquidity;
+        uint24 fee;
+    }
+
+    struct PositionParams {
+        uint160 sqrtRatioX96Lower;
+        uint160 sqrtRatioX96Upper;
+        uint256 amount0Initial;
+        uint256 amount1Initial;
+    }
+
+    function calculateConstantLiquidityPostSwapSqrtPrice(
+        PoolParams memory poolParams,
+        PositionParams memory positionParms
+    ) internal pure returns (uint160 postSwapSqrtRatioX96) {
+        // given constant liquidty / current price / bounds / initialAmounts - calculate how much the price should move
+        // so that the token ratios are of equal liquidity.
+        // will switch from quadratic to binary search
+    }
+
+    function swapToNextInitializedTick(
+        PoolParams memory poolParams,
+        PositionParams memory positionParams,
+        uint160 sqrtRatioX96Target,
+        bool zeroForOne
+    )
+        internal
+        view
+        returns (
+            bool doSwap,
+            uint256 amount0Updated,
+            uint256 amount1Updated
+        )
+    {
+        if (zeroForOne) {
+            if (sqrtRatioX96Target < positionParams.sqrtRatioX96Lower) {
+                return (false, 0, 0);
+            }
+        } else {
+            if (sqrtRatioX96Target > positionParams.sqrtRatioX96Upper) {
+                return (false, 0, 0);
+            }
+        }
+
+        uint256 token0Delta =
+            SqrtPriceMath.getAmount0Delta(
+                poolParams.sqrtRatioX96,
+                sqrtRatioX96Target,
+                poolParams.liquidity,
+                zeroForOne ? false : true
+            );
+        uint256 token1Delta =
+            SqrtPriceMath.getAmount1Delta(
+                poolParams.sqrtRatioX96,
+                sqrtRatioX96Target,
+                poolParams.liquidity,
+                zeroForOne ? true : false
+            );
+
+        uint256 validDeposit0 =
+            SqrtPriceMath.getAmount0Delta(
+                sqrtRatioX96Target,
+                positionParams.sqrtRatioX96Upper,
+                poolParams.liquidity,
+                false
+            );
+        uint256 validDeposit1 =
+            SqrtPriceMath.getAmount1Delta(
+                sqrtRatioX96Target,
+                positionParams.sqrtRatioX96Lower,
+                poolParams.liquidity,
+                false
+            );
+
+        if (zeroForOne) {
```

Comment:
hmm, i wonder what the gas overhead would be if we just used the quoter here? iuuc this code is basically mocking out the results of a swap?

Reply:
scratch that, for the sake of efficiency we have to manually implement the logic if we only want to run tick transition logic once per function call, the quoter runs it every time